File: python/problem_4_median_of_2_arrays.py
# There are two sorted arrays nums1 and nums2 of size m and n respectively.
# Find the median of the two sorted arrays. The overall run time complexity should be O(log (m+n)).

class Solution(object):
    def findMedianSortedArrays(self, nums1, nums2):
        len1, len2 = len(nums1), len(nums2)
        i, j, preLast, last = 0, 0, 0, 0
        sum = len1 + len2
        stopAt = int(sum / 2) + 1
        for x in range(0, stopAt):
            if len1 > i and (len2 <= j or nums1[i] <= nums2[j]):
                preLast = last
                last = nums1[i]
                i = i + 1
            else:
                preLast = last
                last = nums2[j]
                j = j + 1

        return (last + preLast) / 2.0 if sum % 2 == 0 else float(last)



if __name__ == "__main__":
    solution = Solution()
    print(solution.findMedianSortedArrays([1, 3], [2]))
    print(solution.findMedianSortedArrays([1, 3, 5, 7], [2, 4, 6]))
    print(solution.findMedianSortedArrays([1, 3, 5], [2, 4, 6]))

File: csharp/Problem_4_Median_of_2_arrays.cs
using System;
using System.Diagnostics;

namespace LeetCode
{
    public class Problem_4_Median_of_2_arrays
    {
        public void Run()
        {
            Debug.Assert(FindMedianSortedArrays(new[] { 1, 2 }, new[] { 3, 4 }) == 2.5d);
            Debug.Assert(FindMedianSortedArrays(new[] { 1, 3 }, new[] { 2 }) == 2d);
            Debug.Assert(FindMedianSortedArrays(new[] { 2, 4 }, new int[0]) == 3d);
            Debug.Assert(FindMedianSortedArrays(new[] { 2 }, new int[0]) == 2d);
            Debug.Assert(FindMedianSortedArrays(new int[0], new[] { 2 }) == 2d);
        }

        public double FindMedianSortedArrays(int[] nums1, int[] nums2)
        {
            int i = 0, j = 0, preLast = 0, last = 0;
            int sum = nums1.Length + nums2.Length;
            int stopAt = sum / 2;
            for (int k = 0; k <= stopAt; k++)
            {
                if (nums1.Length > i && (nums2.Length <= j || nums1[i] <= nums2[j]))
                {
                    preLast = last;
                    last = nums1[i];
                    i++;
                }
                else
                {
                    preLast = last;
                    last = nums2[j];
                    j++;
                }
            }

            return sum % 2 == 0 ? (last + preLast) / 2d : last;
        }
    }
}